pkg/deploy/deploy.go

```diff
@@ -158,20 +159,42 @@ func (d *deployer) deploy(ctx context.Context, rgName, deploymentName, vmssName
 	for i := 0; i < 3; i++ {
 		d.log.Printf("deploying %s", deploymentName)
 		err = d.deployments.CreateOrUpdateAndWait(ctx, rgName, deploymentName, deployment)
-		if serviceErr, ok := err.(*azure.ServiceError); ok &&
-			serviceErr.Code == "DeploymentFailed" &&
-			i < 1 {
-			// on new RP deployments, we get a spurious DeploymentFailed error
-			// from the Microsoft.Insights/metricAlerts resources indicating
-			// that rp-lb can't be found, even though it exists and the
-			// resources correctly have a dependsOn stanza referring to it.
-			// Retry once.
-			d.log.Print(err)
-			continue
+		serviceErr, isServiceError := err.(*azure.ServiceError)
+
+		// Check for two known errors that we know how to handle.
+		if isServiceError && serviceErr.Code == "DeploymentFailed" && len(serviceErr.Details) > 0 {
+			outerCode, outerCodeOk := serviceErr.Details[0]["code"].(string)
+			message, messageOk := serviceErr.Details[0]["message"].(string)
+
+			if outerCodeOk && messageOk && outerCode == "BadRequest" {
+				innerErr := map[string]interface{}{}
+
+				if jsonErr := json.Unmarshal([]byte(message), &innerErr); jsonErr == nil {
+					innerCode, innerCodeOk := innerErr["code"].(string)
+					innerMessage, innerMessageOk := innerErr["message"].(string)
+
+					if i < 1 && innerCodeOk && innerMessageOk && innerCode == "ResourceNotFound" && strings.Contains(innerMessage, "Microsoft.Network/loadBalancers/rp-lb") {
+						// on new RP deployments, we get a spurious DeploymentFailed error
+						// from the Microsoft.Insights/metricAlerts resources indicating
+						// that rp-lb can't be found, even though it exists and the
+						// resources correctly have a dependsOn stanza referring to it.
+						// Retry once.
+						d.log.Print(err)
+						continue
+					} else if innerCodeOk && innerCode == "CannotModifyProbeUsedByVMSS" {
+						// removes the probe reference so we can update the lb rules
+						if retry := d.vmssCleaner.UpdateVMSSProbes(ctx, rgName); retry {
+							continue
+						}
+					}
+				}
+			}
 		}
+
+		// For errors we don't know how to handle, delete the failed VMSS and retry the deployment.
 		if err != nil && *d.config.Configuration.VMSSCleanupEnabled {
 			if retry := d.vmssCleaner.RemoveFailedNewScaleset(ctx, rgName, vmssName); retry {
-				continue // Retry deployment after deleting failed VMSS.
+				continue
```

Comment:
why is the comment removed here, is it not applicable anymore?

Reply:
I reworded it for a bit more clarity and moved it to be above the outer if block 2 lines up from here.